File: frontend/src/types/job.ts
export type JobStatus =
  | "queued"
  | "running"
  | "succeeded"
  | "failed"
  | "cancelled"
  | "interrupted";

export interface Job {
  id: number;
  job_type: string;
  title: string;
  status: JobStatus;
  dataset_id: number | null;
  stage: string;
  progress_current: number;
  progress_total: number | null;
  error_count: number;
  attempt: number;
  parameters: Record<string, unknown>;
  result: Record<string, unknown> | null;
  error: Record<string, unknown> | null;
  retry_of_id: number | null;
  cancel_requested_at: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  updated_at: string;
}

export interface JobListResponse {
  items: Job[];
  total: number;
}

export interface ScanJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface AnnotationExportJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface MetadataImportJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface MetadataImportRollbackJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface LabelmeImportJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface LabelmeImportRollbackJobCreateResponse {
  job: Job;
  created: boolean;
}

export interface ThumbnailJobCreateResponse {
  job: Job | null;
  created: boolean;
  requested_count: number;
  eligible_count: number;
  cached_count: number;
  skipped_count: number;
}

export interface ThumbnailMaintenanceJobCreateResponse {
  job: Job | null;
  created: boolean;
  due: boolean;
}

export interface DatabaseBackupJobCreateResponse {
  job: Job;
  created: boolean;
}

File: backend/app/schemas/job.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


JobStatus = Literal[
    "queued",
    "running",
    "succeeded",
    "failed",
    "cancelled",
    "interrupted",
]


class JobRead(BaseModel):
    id: int
    job_type: str
    title: str
    status: JobStatus
    dataset_id: int | None
    stage: str
    progress_current: int
    progress_total: int | None
    error_count: int
    attempt: int
    parameters: dict[str, object]
    result: dict[str, object] | None
    error: dict[str, object] | None
    retry_of_id: int | None
    cancel_requested_at: datetime | None
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    updated_at: datetime


class JobListResponse(BaseModel):
    items: list[JobRead]
    total: int


class JobCreate(BaseModel):
    job_type: str = Field(min_length=1, max_length=80)
    title: str = Field(min_length=1, max_length=200)
    dataset_id: int | None = None
    parameters: dict[str, object] = Field(default_factory=dict)
    progress_total: int | None = Field(default=None, ge=0)
